make event dispatching non-blocking

benjamin-stacks opened the pull request:
addresses #6543

### Checklist

- [x] Test coverage for new or modified code paths
- [x] Changelog is updated
- [x] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [x] ~New clarity functions have corresponding PR in `clarity-benchmarking` repo~

benjamin-stacks commented on stacks-node/src/event_dispatcher/db.rs:
```diff
@@ -0,0 +1,322 @@
+use std::path::PathBuf;
```
Ah, good call, will do. Is "Stacks Open Internet Foundation" still the correct copyright holder?

Also, there's a whole bunch of files that lack that header, I'm wondering if there's a good way to automate this.

benjamin-stacks committed:
feat: non-blocking event delivery

This commit is the main implementation work for #6543. It moves event
dispatcher HTTP requests to a separate thread. That way, a slow event
observer doesn't block the node from continuing its work.

Only if your event observers are so slow that the node is continuously
producing events faster than they can be delivered, will it eventually
start blocking again, because the queue size for pending requests is
bounded (at 1,000 right now, but I picked that number out of a hat,
happy to change it if anyone has thoughts).

Each new event payload is stored in the event observer DB, and its
ID is then sent to the subthread, which will make the request and then
delete the DB entry.

That way, if a node is shut down while there are pending requests,
they're in the DB ready to be retried after restart via
`process_pending_payloads()` (which blocks until completion). So that's
exactly as before (except that previously there couldn't have been more
than one or two pending payloads).

benjamin-stacks committed:
fix: add a way to block until all events are delivered, to fix a test

This fixes [this integration test failure](https://github.com/stacks-network/stacks-core/actions/runs/20749024845/job/59577684952?pr=6762),
caused by the fact that event delivery wasn't complete by the time the
assertions were made.

benjamin-stacks committed:
fix: ensure `process_pending_payloads` is really only called at startup

Doing this work in the RunLoop implementations' startup code is *almost*
the same thing, but not quite, since the nakamoto run loop might be
started later (after an epoch 3 transition), at which point the event DB
may already have new items from the current run of the application,
which should *not* be touched by `process_pending_payloads`.

This used to not be a problem, but now that that DB is used for the
actual queue of the (concurrently running) EventDispatcherWorker, it has
become one.

benjamin-stacks committed:
fix all the integration test issues with asynchronous event dispatching

This is like 72437b24db750af4bb26ddb56361a2c26cd063d1, but it works for
all the tests instead of only the one.

While only that one test very obviously failed, the issue exists for
pretty much all of the integration tests, because they rely on the
test_observer to capture all relevant data.

Things are fast enough, and therefore we've only seen one blatant
failure, but

1) it's going to be flaky (I can create a whole lot of test failures
   by adding a small artificial delay to event delivery), and
2) it might actually be *hiding* test failures (in some cases, like e.g.
   neon_integrations::deep_contract, we're asserting that certain things
   are *not* in the data, and if the data is incomplete to begin with,
   those assertions are moot).

benjamin-stacks committed:
use the same event dispatcher for neon and nakamoto

When switching runloops at the epoch 2/3 transition, this ensures that
the same event dispatcher worker thread is handling delivery, which in
turn ensures that all payloads are delivered in order

benjamin-stacks committed:
give event dispatcher threads distinct names, and at some debug logging

benjamin-stacks committed:
Merge branch 'refactor/event-dispatcher-tweaks' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'refactor/event-dispatcher-tweaks' into feat/non-blocking-event-delivery

benjamin-stacks committed:
refactor: remove error boilerplate by using `thiserror`

Thanks Hank for the tip!

benjamin-stacks committed:
remove unused import

Not sure why this wasn't caught in the pre-commit hook, I'd have assumed
the checks are the same.

benjamin-stacks committed:
Merge branch 'refactor/event-dispatcher-tweaks' into feat/non-blocking-event-delivery

benjamin-stacks committed:
feat: allow specifying the queue size for the event dispatcher

The logic is slightly tricky here, because the size of the queue (the
max number of in-flight requests before we start blocking the thread) is
implemented through the `bound` parameter of the `sync_channel`, but
those two values aren't actually the same.

See the comment at the top of `EventDispatcherWorker::new()` for
details.

benjamin-stacks committed:
allow configuring event dispatcher blocking behavior in `config.toml`

See the discussion in https://github.com/stacks-network/stacks-core/issues/6543
for some background.

benjamin-stacks commented on stackslib/src/config/mod.rs:
```diff
@@ -2145,6 +2145,30 @@ pub struct NodeConfig {
+    /// to `true`, as no in-flight requests are allowed.
+    /// ---
+    /// @default: `1_000`
+    pub event_dispatcher_queue_size: usize,
```
This value ultimately ends up as the `bound` argument to [`sync_channel`](https://doc.rust-lang.org/std/sync/mpsc/fn.sync_channel.html), which is of type `usize`.

I don't know if we have any concerns about having a platform-dependent number type on the configuration object -- if yes, we can also use something else here. Arguably, any *reasonable* values for this setting should fit into 16 bits anyway. If you need your queue to be bigger than 64k, you should reconsider you architecture.

benjamin-stacks committed:
chore: update changelog

benjamin-stacks committed:
chore: add/bump copyright note in all files that this PR touches

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
tweak so that WorkerTask::NoOp doesn't exist in all during non-test runs

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2783820525

benjamin-stacks committed:
the attributes need to match the usage elsewhere

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

jcnelson commented on stacks-node/src/event_dispatcher/tests.rs:
```diff
@@ -970,3 +990,253 @@ fn test_block_proposal_validation_event() {
+        "dispatcher did not block while sending event"
+    );
+
+    thread::sleep(Duration::from_millis(100));
```
Flagging this as well, to try and use something more robust to verify that the event dispatcher is making progress correctly.

benjamin-stacks replied:
responded [here](https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2834712493)

jcnelson commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,397 @@
+            return;
+        }
+
+        // We're deleting regardless of result -- if retries are disabled, that means
```
Please instead keep the payload in the DB until the worker gets an HTTP 200.  The node or event observer can crash at any time (for example, due to OOM or file descriptor exhaustion on misconfigured hosts), and we'd want the node at least to be robust in the face of this outcome (i.e. by re-trying any inflight but undelivered messages).

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,397 @@
+                    // If the sending fails (i.e. the receiver has been dropped), that means a logic bug
+                    // has been introduced to the code -- at time of writing, the main function is waiting
+                    // for this message a few lines down, outside the thread closure.
+                    // We log this, but we still start the loop.
```
This is about the channel over which the worker thread reports to the main thread that it's ready, not about communication *to* the worker.

benjamin-stacks replied:
Yeah, that's fair. I'll change this to panic instead of just logging. Our panic hook will then kill the whole node.

jcnelson commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,397 @@
+        }
+
+        // We're deleting regardless of result -- if retries are disabled, that means
+        // we're supposed to forget about it in case of failure. If they're not disabled,
```
Retries are not supposed to be disable-able.  The node guarantees at-least-once delivery of events to all downstream event observers.

jcnelson commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,397 @@
+        // Cap the backoff at 3x the timeout
+        let max_backoff = data.timeout.saturating_mul(3);
+
+        loop {
```
Could this be factored to use the `with_retry()` function above?

Also, I see that this code performs retry jitter, whereas `with_retry()` does not.  Is there a reason for this?

benjamin-stacks replied:
> Could this be factored to use the `with_retry()` function above?

Possibly, but as you noted, the two behave differently. That is why I didn't use `with_retry` for the HTTP request, because I didn't want to introduce unrelated functional changes or make `with_retry()` more complex to support a single use case.

As for the question *why* that difference is there, I could guess, but I don't know for sure because that code predates me by a long time.

The `with_retry` logic was added in #5358, I just moved it to a helper so I could reuse it for `get_payload_with_retry()`.

The backoff jitter for the HTTP request retries was added in #5327.

Since both of those came from @brice-stacks, I guess he could add some color here, but in either case, I don't think a change to any of this needs to be in this PR.

jcnelson reviewed (changes requested):
This overall looks good, but I have a few concerns about the threadpool and the handling of clock skew (among other lesser things).

benjamin-stacks committed:
remove catch_up_all_event_dispatchers() and supporting infrastructure

as [Jude points out](https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2834777416),
it's no longer necessary since the default behavior is still blocking

benjamin-stacks committed:
if the dispatcher worker receives an invalid payload, panic instead of spinning forever

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2834737738

benjamin-stacks committed:
panic instead of just logging if the thread can't report being ready

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2834739553

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

jcnelson reviewed (dismissed):
LGTM, just one comment with an easy resolution

benjamin-stacks committed:
remove a level of nesting from the "event is old" check

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r2992190808

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,397 @@
+            // is *after* `now` (which should be extremely rare), the most likely reason is a *slight*
+            // adjustment to the the system clock (e.g. NTP sync) that happened between storing the
+            // entity and retrieving it, and that should be fine.
+            // If there was a *major* adjustment, all bets are off anyway. You shouldn't mess with your
```
Ok, I think I'll leave the timestamp as-is, because in my mind, the case where a payload lives across restarts (and thus can't be monotonic because it comes from the DB) may be the most relevant to handle, as that's most likely to be dispatched with a delay.

Agree with removing a level of nesting, although I can't do `saturating_add` since this is a typed timestamp (and I have a strong preference for keeping it that way and not passing around untyped numbers). But it'll take about 292 billion years for this to panic (and it doesn't involve external data and thus isn't an attack vector), so I'm okay with it.

That's done in ae3e5e56ec6a9667e521b9c13a6711770994001e.

benjamin-stacks committed:
migrate changelog entry to the new system

benjamin-stacks committed:
use `from_secs` instead of `from_mins` because the former is unstable pre-1.91

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
use correct queue size when starting a new dispatcher in the naka run loop

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r3010024358

benjamin-stacks committed:
fix typypo

Co-authored-by: Brice <232827048+brice-stacks@users.noreply.github.com>

benjamin-stacks committed:
Merge remote-tracking branch 'origin-benjamin/feat/non-blocking-event-delivery' into feat/non-blocking-event-delivery

brice-stacks commented on stacks-node/src/event_dispatcher/db.rs:
```diff
@@ -267,8 +247,78 @@ impl EventDispatcherDbConnection {
+/// # Example
+///
+///     let response = with_retry(|| perform_db_op(42), "database operation 42 failed");
+fn with_retry<T, E, F>(f: F, error_log_text: String) -> T
```
Does this need to take a `String`? Could it just take a `&str`?

benjamin-stacks replied:
Indeed, good call. Changed in 31cf1964f4481a45ccc7eb52d47595760dbcc7ee.

brice-stacks reviewed (dismissed):
This looks good to me, I just left one possible small suggestion.

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks committed:
no need to create `String`s for `with_retry` fail messages

https://github.com/stacks-network/stacks-core/pull/6762#discussion_r3047214125

brice-stacks reviewed (dismissed):
lgtm!

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

brice-stacks reviewed (dismissed):
lgtm!

dhaney-stacks commented:
@benjamin-stacks please remove @jcnelson as reviewer and add anyone else as appropriate. Or merge this!

benjamin-stacks commented:
> @benjamin-stacks please remove @jcnelson as reviewer and add anyone else as appropriate. Or merge this!

@dhaney-stacks I have removed the review request, but I can't remove him from the list of reviewers because he's left comments and reviews already, and so in GitHub's book he counts as a reviewer.

Also I'd love to merge this but it has only one approval.

francesco-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+        }
+
+        #[cfg(test)]
+        if TEST_EVENT_OBSERVER_SKIP_RETRY.get() {
```
Small test-isolation issue here: `TEST_EVENT_OBSERVER_SKIP_RETRY` is process-global, and this PR makes it much easier for that state to leak across tests because delivery now happens asynchronously. I was able to get `test_send_payload_retry` to fail while the worker logged `Fault injection: skipping retry of payload` and `Fault injection: skipping deletion of payload`, even though that test never sets the flag itself. It looks like `test_send_payload_with_db_force_restart` can temporarily change the flag while unrelated work is still in flight.

A possible fix could be to move this `TEST_EVENT_OBSERVER_SKIP_RETRY` to the dispatcher or worker instance under `#[cfg(test)]` instead of keeping it global

francesco-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+                        return Err(err.into());
+                    }
+                    #[cfg(test)]
+                    if TEST_EVENT_OBSERVER_SKIP_RETRY.get() {
```
same issue as above

francesco-stacks commented on stacks-node/src/event_dispatcher/tests.rs:
```diff
@@ -1076,3 +1096,253 @@ fn test_block_proposal_validation_event() {
+    );
+
+    assert!(
+        start.elapsed() < Duration::from_millis(100),
```
This seems to be flaky on my end (~1/15 though) could it make sense to increase this a bit?

benjamin-stacks replied:
Interesting. In all the months this PR has lived, I don't recall any of this ever flaking even once, either in CI nor locally. Can you give me some details on how you ran the tests?

As mentioned elsewhere, I chose these values pretty high, so the behavior would for all intents and purposes be deterministic. I'd love to understand where the assumptions are incorrect.

francesco-stacks replied:
I just launched `for i in {1..50}; do cargo test -p stacks-node -- event_dispatcher; done` while also using the computer

francesco-stacks replied:
in the interest of not keeping this blocked forever just for some tets, I'am going to resolve this issue, if and when we'll see some flaky tests on CI will fix them

benjamin-stacks commented on stacks-node/src/event_dispatcher.rs:
```diff
@@ -237,7 +211,9 @@ impl ProposalCallbackReceiver for ProposalCallbackHandler {
 
         for observer in self.observers.iter() {
             self.dispatcher
-                .dispatch_to_observer(observer, &response, PATH_PROPOSAL_RESPONSE);
+                .dispatch_to_observer(observer, &response, PATH_PROPOSAL_RESPONSE)
+                .unwrap()
+                .wait_until_complete();
         }
```
> If enqueueing fails (e.g., worker thread/channel closed), this will panic and can take down the process.

Indeed, and it should.

benjamin-stacks commented on stacks-node/src/event_dispatcher/tests.rs:
```diff
@@ -1076,3 +1096,253 @@ fn test_block_proposal_validation_event() {
+        "dispatcher blocked while sending first three events"
+    );
+
+    thread::sleep(Duration::from_millis(500) - elapsed);
```
Half a second is an eternity in practice. I'd rather look at actual problems (like the one @francesco-stacks mentions elsewhere) than theoretical ones.

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+                // this will run forever until the messaging channel is closed
+                Self::main_thread_loop(conn, message_rx);
+            })
+            .unwrap();
```
Not being able to spawn the worker thread isn't a recoverable error. If we're looking for consistency, we might consider panicking in all cases here, but that feels more like a stylistic question. In either case, re this:

> so callers can handle startup failures cleanly

the only caller of that function `expect`s this result, so it's gonna panic either way. 🤷

benjamin-stacks committed:
fix a couple of typos pointed out by the robot

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery

benjamin-stacks commented on stacks-node/src/event_dispatcher/db.rs:
```diff
@@ -68,33 +70,21 @@ impl EventDispatcherDbConnection {
+                Err(db_error::SqliteError(rusqlite::Error::QueryReturnedNoRows)) => Ok(None),
+                Err(e) => Err(e),
+            },
+            "Failed to retrieve payload {id} from event observer database",
```
oops, indeed

fixed in a6d4edad4104e0f04420ee15f4f289e2ab5c7c2d

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+        let Some(receiver) = self.receiver else {
+            return;
+        };
+        // There is no codepath that would drop the sender without sending the acknowledgenent
```
a6d4edad4104e0f04420ee15f4f289e2ab5c7c2d

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+        // In other words, the channel's bound has to be one less than
+        // the desired queue size. If the queue size is 0, then the bound
+        // is also set to zero, and *in addition* we block until the
+        // request has happend.
```
a6d4edad4104e0f04420ee15f4f289e2ab5c7c2d

benjamin-stacks commented on stacks-node/src/event_dispatcher/worker.rs:
```diff
@@ -0,0 +1,467 @@
+        })?;
+
+        // we're ignoring `must_block` here -- the only point of `noop()` is to call
+        // `wait_until_complete()` immedately anyway, so we'll let the caller do that.
```
a6d4edad4104e0f04420ee15f4f289e2ab5c7c2d

benjamin-stacks commented on stacks-node/src/main.rs:
```diff
@@ -265,6 +281,20 @@ fn cli_get_miner_spend(
+/// Pending events for observers that are no longer registered will be discarded.
+fn send_pending_event_payloads(conf: &Config) {
+    // This dispatcher gets a queue size of 0 to ensure that it blocks. Technically
+    // process_pending_payloads() always blocks; this is just and additional safeguard.
```
a6d4edad4104e0f04420ee15f4f289e2ab5c7c2d

francesco-stacks reviewed (approved):
changes looks good to me! there are some nits regarding the tests that we'll revisit if they will ever cause a problem. but not worth to keep this pr blocked for those

benjamin-stacks committed:
Merge branch 'develop' into feat/non-blocking-event-delivery
